src/main/java/org/reactome/addlinks/dataretrieval/ensembl/EnsemblBiomartRetriever.java

```diff
@@ -0,0 +1,232 @@
+package org.reactome.addlinks.dataretrieval.ensembl;
+
+import org.apache.http.HttpException;
+import org.apache.http.HttpStatus;
+import org.json.simple.parser.ParseException;
+import org.reactome.addlinks.EnsemblBiomartUtil;
+import org.reactome.addlinks.dataretrieval.FileRetriever;
+
+import java.io.*;
+import java.net.HttpURLConnection;
+import java.net.URL;
+import java.net.URLEncoder;
+import java.nio.file.Files;
+import java.nio.file.Paths;
+import java.nio.file.StandardOpenOption;
+import java.time.Duration;
+import java.util.*;
+
+public class EnsemblBiomartRetriever extends FileRetriever {
+
+    private Properties properties;
+
+    public EnsemblBiomartRetriever() { }
+
+    public EnsemblBiomartRetriever(String retrieverName)
+    {
+        super(retrieverName);
+    }
+
+    private static final String  microarrayTypesBaseQuery = "type=listAttributes&mart=ENSEMBL_MART_ENSEMBL&virtualSchema=default&dataset=BIOMART_SPECIES_NAME_gene_ensembl&interface=default&attributePage=feature_page&attributeGroup=external&attributeCollection=microarray";
+
+    /**
+     * Downloads Ensembl-Microarray and Ensembl-Uniprot identifier mapping files for all species, if they exist.
+     * @throws IOException - Thrown when file can't be found when during writing and by HTTPConnection class.
+     * @throws ParseException - Thrown by JSONParser when reading Species config file.
+     * @throws InterruptedException - Thrown if Sleep is interrupted when waiting to retry BioMart query.
+     * @throws BioMartQueryException - Thrown if BioMart query doesn't match any existing data in their database.
+     * @throws HttpException - Thrown when the Http request to BioMart returns a non-200 response.
+     */
+    public void downloadData() throws IOException, ParseException, InterruptedException, BioMartQueryException, HttpException {
+
+        // Create directory where Biomart files will be stored.
+        Files.createDirectories(Paths.get(this.destination));
+        setProperties(EnsemblBiomartUtil.getProperties());
+
+        // Get names of all organisms we add links and/or microarray data for.
+        // Species names are in biomart format (eg: hsapiens).
+        for (String speciesName : EnsemblBiomartUtil.getSpeciesNames()) {
+            String speciesBiomartName = EnsemblBiomartUtil.getBiomartSpeciesName(speciesName);
+            logger.info("Retrieving Biomart files for " + speciesBiomartName);
+
+            logger.info("Retrieving microarray data");
+            // Query Biomart for existing microarray 'types' (not ids) that exist for this species.
+            Set<String> microarrayTypes = queryBiomart(getMicroarrayTypesQuery(speciesBiomartName), 0);
```

Comment:
Turning the 0 into a named constant at the start of the method such as `int initialRetryCount = 0` would help give context without needing to go to the method signature of `queryBiomart` to understand its significance.

Reply:
Done